cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java

```diff
@@ -0,0 +1,469 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.storage.spanner;
+
+import com.google.cloud.ByteArray;
+import com.google.cloud.spanner.Key;
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.TransactionContext;
+import com.google.cloud.spanner.Value;
+import io.cdap.cdap.api.dataset.lib.AbstractCloseableIterator;
+import io.cdap.cdap.api.dataset.lib.CloseableIterator;
+import io.cdap.cdap.spi.data.InvalidFieldException;
+import io.cdap.cdap.spi.data.StructuredRow;
+import io.cdap.cdap.spi.data.StructuredTable;
+import io.cdap.cdap.spi.data.table.StructuredTableSchema;
+import io.cdap.cdap.spi.data.table.field.Field;
+import io.cdap.cdap.spi.data.table.field.FieldType;
+import io.cdap.cdap.spi.data.table.field.FieldValidator;
+import io.cdap.cdap.spi.data.table.field.Fields;
+import io.cdap.cdap.spi.data.table.field.Range;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link StructuredTable} implementation backed by GCP Cloud Spanner.
+ */
+public class SpannerStructuredTable implements StructuredTable {
```

Comment:
A high level question/decision is: using DML or Mutation
https://cloud.google.com/spanner/docs/dml-versus-mutations#what_is_data_manipulation_language_dml_

Key deltas are: 
1. DML supports reading uncommitted results within an active transaction, but not mutation, which is applied atomically. 
2. DML does NOT support upsert, but mutation does. 

Also note that mutation is a native way cloudspanner process writes in the backend, so potentially more performant. 

Not certain which is more frequent: reading uncommited results  or upsert. Suspect the latter. 

Fine with leaving this as is. We can switch later if needed.

Reply:
To clarify: switch only upsert operation to using mutation. Other operations don't look to benefit from mutation.

Reply:
The upsert also seems to be a lazy design. A lot of usages of `upsert` can be replaced with either `insert` or `update` since we actually know in most cases.

Reply:
Also, using Mutation for upsert can easily creates surprises at the StructuredTable level, which is supposed to support read your own writes in the same TX.

Reply:
Ack. As discussed offline, long term goal is to get rid of upsert